experimental/Matrix/matrix.g

```diff
@@ -0,0 +1,192 @@
+
+InstallOtherMethod(\+, [IsJuliaMatrixRep, IsJuliaMatrixRep], function( m1, m2 )
+        return MakeJuliaMatrixRep(m1!.m + m2!.m);
+    end);
+    
+InstallOtherMethod(\-, [IsJuliaMatrixRep, IsJuliaMatrixRep], function( m1, m2 )
+        return MakeJuliaMatrixRep(m1!.m - m2!.m);
+    end);
+
+InstallOtherMethod(\^, [IsJuliaMatrixRep, IsInt and IsSmallIntRep], function( m1, k )
+        if k < 0 then
+            return MakeJuliaMatrixRep( (Inverse(m1)!.m)^(-k) );
+        else
+            return MakeJuliaMatrixRep( (m1!.m)^k );
+        fi;
+    end);
+    
+InstallOtherMethod(Display, [IsJuliaMatrixRep], function( m )
+    local i,j;
+        Print("[ ");
+        for i in [1..NumberRows(m)] do
+            Print("[ ");
+            for j in [1..NumberColumns(m)] do
+                Print(m[i,j]);
+                if (j < NumberColumns(m)) then
+                    Print(", ");
+                fi;
+            od;
+            if (i < NumberRows(m)) then
+                Print(" ] \n");
+                Print("  ");
+            else
+                Print(" ] ] \n");
+            fi;
+        od;
+    end);
+
+InstallOtherMethod(Unpack, [IsJuliaMatrixRep], function(m)
+    local v, i, j;
+    v := EmptyPlist(NumberRows(m)*NumberColumns(m));
+    for i in [1..NumberRows(m)] do
+        for j in [1..NumberColumns(m)] do
+            v[j+NumberColumns(m)*(i-1)] := m[i,j];
+        od;
+    od;
+    return v;
+end);
+
+
+# Method: SetNiceMorphismForJuliaMatrixRepGroup
+# Input: Matrix group G
+# Output: -
+# This function sets a nice morphism for G. The morphism is defined from the Julia matrix to the corresponding GAP matrix
+
+DeclareOperation("SetNiceMorphismForJuliaMatrixRepGroup", [IsGroup]);
+
+InstallMethod(SetNiceMorphismForJuliaMatrixRepGroup, [IsGroup], function(G)
+    local hom, f, f_inv, ele, GAPGenerators, i, gens, GAPGroup, JuliaGAPMap;
+    
+        gens := GeneratorsOfGroup(G);
+        Assert(0, ForAll(gens, IsJuliaMatrixRep));
+        
+        ele := gens[1];
+        hom := Julia.Oscar._iso_oscar_gap(ele!.m.base_ring);
+        
+        GAPGenerators := List(gens, g -> Julia.AbstractAlgebra.map_entries(hom, g!.m));
+        
+        GAPGroup := GroupByGenerators(GAPGenerators);
+        
+        f := function(m) return Julia.AbstractAlgebra.map_entries(hom,m!.m); end;
+        f_inv := function(m) return MakeJuliaMatrixRep(Julia.Oscar.preimage_matrix(hom,m)); end;
+
+        JuliaGAPMap := GroupHomomorphismByFunction(G,GAPGroup,f,f_inv);
+        SetNiceMonomorphism(G,JuliaGAPMap);
+        SetIsHandledByNiceMonomorphism(G, true);
+    end);
+    
+InstallOtherMethod(\=, [IsJuliaMatrixRep, IsJuliaMatrixRep], function( m1, m2 )
+        return m1!.m = m2!.m;
+    end);
+    
+# TODO: Not working in all cases. Needs more checks?
+InstallOtherMethod(\<, [IsJuliaMatrixRep, IsJuliaMatrixRep], function( m1, m2 )
+        return m1!.m < m2!.m;
+    end);
+
```

Comment:
It won't work for almost all Oscar matrices. I see two options:
1. We simply insert something like `Error("comparing IsJuliaMatrixRep not supported");` -> easy to do, will allow us quickly to identify places where it is needed
2. Implement a lexicographical comparison ourselves (ideally as a Julia function, to make it faster). I.e., write a little Julia helper like this:
```julia
function _lex_isless(a::T,b::T) where T<:MatElem
  @assert base_ring(a) === base_ring(b)
  @assert size(a) == size(b)
  for i in 1:nrows(a), j in 1:ncols(a)
    if a[i,j] != b[i,j]
      return a[i,j] < b[i,j]
    end
  end
  return false
end
```
Alas, that will only work for base rings QQ and ZZ. For finite field or number fields it'll fail, because those can't be compared. We could fudge something for finite fields... but I really think we should first follow option 1 and see where it gets us.

Reply:
I tried to merge both ideas. What do you think about @fingolfin ?